File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/simple/BoxLineReduction.kt
package sudokusolver.kotlin.logic.simple

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.mergeToRemoveCandidates

/*
 * http://www.sudokuwiki.org/Intersection_Removal#LBR
 *
 * For a given row, if a candidate appears in only one block, then the candidate for that block must be placed in that
 * row. The candidate can be removed from the cells which are in the same block, but different rows.
 *
 * For a given column, if a candidate appears in only one block, then the candidate for that block must be placed in
 * that column. The candidate can be removed from cells which are in the same block, but different columns.
 */
fun boxLineReduction(board: Board<Cell>): List<RemoveCandidates> =
    SudokuNumber.entries.flatMap { candidate ->

        fun boxLineReduction(units: List<List<Cell>>, getUnitIndex: (Cell) -> Int) =
            units.mapNotNull { unit ->
                unit.asSequence()
                    .filterIsInstance<UnsolvedCell>()
                    .filter { candidate in it.candidates }
                    .map { it.block }
                    .toSet()
                    .singleOrNull()
                    ?.let { blockIndex ->
                        val unitIndex = getUnitIndex(unit.first())
                        board.getBlock(blockIndex)
                            .filterIsInstance<UnsolvedCell>()
                            .filter { getUnitIndex(it) != unitIndex && candidate in it.candidates }
                            .map { it to candidate }
                    }
            }.flatten()

        val rowRemovals = boxLineReduction(board.rows) { it.row }
        val columnRemovals = boxLineReduction(board.columns) { it.column }
        rowRemovals + columnRemovals
    }.mergeToRemoveCandidates()

File: SudokuSolver_Swift/SudokuSolver_Swift/logic/simple/BoxLineReduction.swift
/*
 * http://www.sudokuwiki.org/Intersection_Removal#LBR
 *
 * For a given row, if a candidate appears in only one block, then the candidate for that block must be placed in that
 * row. The candidate can be removed from the cells which are in the same block, but different rows.
 *
 * For a given column, if a candidate appears in only one block, then the candidate for that block must be placed in
 * that column. The candidate can be removed from cells which are in the same block, but different columns.
 */
func boxLineReduction(board: Board<Cell>) -> [BoardModification] {
    SudokuNumber.allCases.flatMap { candidate -> [LocatedCandidate] in
        
        func boxLineReduction(units: [[Cell]], getUnitIndex: (Location) -> Int) -> [LocatedCandidate] {
            units.flatMap { unit in
                let blockIndices = Set(unit.unsolvedCells.filter { $0.candidates.contains(candidate) }.map(\.block))
                if let blockIndex = blockIndices.first, blockIndices.count == 1 {
                    let unitIndex = getUnitIndex(unit.first!)
                    return board.getBlock(blockIndex: blockIndex)
                        .unsolvedCells
                        .filter { getUnitIndex($0) != unitIndex && $0.candidates.contains(candidate) }
                        .map { ($0, candidate) }
                } else {
                    return []
                }
            }
        }
        
        let rowRemovals = boxLineReduction(units: board.rows, getUnitIndex: \.row)
        let columnRemovals = boxLineReduction(units: board.columns, getUnitIndex: \.column)
        return rowRemovals + columnRemovals
    }.mergeToRemoveCandidates()
}
